Cross-worktree PCH hits with path_remap=off serve stale absolute paths, causing compiler diagnostics + downstream DLL load failures (Windows)

## Summary

When a user has multiple sibling clones of the same project (`C:\Users\niteris\dev\fastled`, `\fastled2`, … `\fastled10`) and builds in one of them, zccache returns HIT verdicts for **PCH files** built in a *different* worktree, even though `path_remap=off`. The served PCH contains absolute paths from the original worktree (`fastled10`), which then propagates into:

1. Compiler diagnostics referencing the wrong worktree paths (cosmetic, but extremely confusing).
2. Suspected downstream DLL build/load corruption on Windows — a deterministic set of tests fail to load with `LoadLibrary` error 126 (`ERROR_MOD_NOT_FOUND`).

This makes parallel-checkout workflows (e.g. for branch-comparison or A/B builds) unreliable on Windows.

## Reproduction

1. Have two sibling worktrees of the same project on Windows: `C:\Users\niteris\dev\fastled7` and `C:\Users\niteris\dev\fastled10`.
2. Build in `fastled10` so PCH `FastLED.h.pch` and `test_pch.h.pch` are populated in zccache.
3. Switch to `fastled7` (same SHA or close), trigger a clean rebuild: `bash test --cpp --clean`.

## Observed (zccache session log, abridged)

```
[DIAG] depgraph_check: C:\Users\niteris\dev\fastled7\src\FastLED.h
  -> C:\Users\niteris\dev\fastled7\.build\meson-quick\ci\meson\native\FastLED.h.pch
  ctx=141f9bb3 verdict=Hit reason=hit: artifact_key=e6c84a2f
  key_root=C:\Users\niteris\dev\fastled7 path_remap=off
[HIT] C:\Users\niteris\dev\fastled7\src\FastLED.h
  -> C:\Users\niteris\dev\fastled7\.build\meson-quick\ci\meson\native\FastLED.h.pch
```

…and downstream compiler output for the same build session:

```
In file included from ../../src/fl/build/fl.stl+.cpp:10:
In file included from C:/Users/niteris/dev/fastled10/src\fl/stl/_build.cpp.hpp:21:
C:/Users/niteris/dev/fastled10/src\fl/stl/not_null.cpp.hpp:24:41: warning: unused parameter 'message'
```

Note: the *source* being compiled is the local `fastled7` checkout (path `../../src/fl/build/fl.stl+.cpp` is relative to the fastled7 build dir), but clang reports the *included* header at a `fastled10` path. That's the cached PCH talking.

## Downstream symptom

Of 304 test DLLs built in this session, 9 fail to load with Windows error 126:

```
Error: Failed to load .../tests/fl_math_fixed_point.dll (error code: 126)
Error: Failed to load .../tests/fl_math_geometry.dll (error code: 126)
Error: Failed to load .../tests/fl_math_lut.dll (error code: 126)
Error: Failed to load .../tests/fl_math_alpha.dll (error code: 126)
Error: Failed to load .../tests/fl_math_ease.dll (error code: 126)
Error: Failed to load .../tests/fl_math_line_simplification.dll (error code: 126)
Error: Failed to load .../tests/fl_math_filter_filter.dll (error code: 126)
Error: Failed to load .../tests/fl_math_math.dll (error code: 126)
Error: Failed to load .../tests/fl_hsv2rgb_accuracy.dll (error code: 126)
```

The set is deterministic across rebuilds. The same SHA passes both Linux and Windows CI on a fresh runner (no parallel worktrees), so the failure is specific to the cross-worktree PCH-serving condition. The `fastled7` rebuild was a `--clean` rebuild — the only persistent state surviving the clean was zccache's cache.

## Why I think this is a zccache bug, not a FastLED bug

- The build invocation references only `fastled7` paths.
- `--clean` blew away `.build/meson-quick/` entirely, but the same `fastled10` paths surface again on rebuild.
- The session log shows `verdict=Hit reason=hit: artifact_key=e6c84a2f path_remap=off` for the PCH. PCH files are *not* relocatable — they bake in absolute header paths. Returning a HIT across worktrees without remapping is incorrect.

## Suggested fixes (any of, ranked)

1. **Treat PCH as worktree-scoped.** Either include `key_root` in the PCH artifact key, or refuse PCH HITs when the served `key_root` does not match the current invocation's.
2. **Default `path_remap=on` for PCH artifacts** even when general object caching has it off. PCH integrity demands the recorded include paths match the include resolution that will happen in the current build.
3. **Emit a louder warning** in `--strict-paths` mode when a HIT is served from a different `key_root` than the current invocation.

## Environment

- Windows 10 Pro 19045 + clang 21.1.5
- zccache 1.11.8 (protocol v14)
- Meson 1.11.1
- Sibling worktrees: 10 parallel clones (`fastled` through `fastled10`).

## Workaround (today)

`zccache clear` plus rebuild — but that defeats the whole point of having the cache. Worktree-isolating PCH artifacts would fix it without disabling caching for the much larger set of non-PCH .obj artifacts.

--- a/crates/zccache/src/daemon/server/handle_compile/hit_branches.rs
+++ b/crates/zccache/src/daemon/server/handle_compile/hit_branches.rs
@@ -118,6 +118,7 @@ pub(super) struct FastHitProbe<'a> {
     pub(super) client_env: Option<&'a [(String, String)]>,
     pub(super) is_rustc: bool,
     pub(super) worktree_equivalent_context: bool,
+    pub(super) worktree_bound: bool,
     pub(super) compile_start: Instant,
     pub(super) parse_args_ns: u64,
     pub(super) build_context_ns: u64,
@@ -139,6 +140,7 @@ pub(super) fn try_fast_hit(probe: FastHitProbe<'_>) -> Option<Response> {
         client_env,
         is_rustc,
         worktree_equivalent_context,
+        worktree_bound,
         compile_start,
         parse_args_ns,
         build_context_ns,
@@ -210,6 +212,7 @@ pub(super) fn try_fast_hit(probe: FastHitProbe<'_>) -> Option<Response> {
             output_path,
             input_paths,
             request_cache_key_root.as_ref(),
+            worktree_bound,
         ),
     );
     Some(response)
@@ -231,6 +234,7 @@ pub(super) struct DepgraphHitProbe<'a> {
     pub(super) client_env: Option<&'a [(String, String)]>,
     pub(super) is_rustc: bool,
     pub(super) worktree_equivalent_context: bool,
+    pub(super) worktree_bound: bool,
     pub(super) compile_start: Instant,
     pub(super) parse_args_ns: u64,
     pub(super) build_context_ns: u64,
@@ -256,6 +260,7 @@ pub(super) fn try_depgraph_cached_hit(probe: DepgraphHitProbe<'_>) -> Option<Res
         client_env,
         is_rustc,
         worktree_equivalent_context,
+        worktree_bound,
         compile_start,
         parse_args_ns,
         build_context_ns,
@@ -325,6 +330,7 @@ pub(super) fn try_depgraph_cached_hit(probe: DepgraphHitProbe<'_>) -> Option<Res
                 output_path,
                 input_paths,
                 request_cache_key_root.as_ref(),
+                worktree_bound,
             ),
         );
     }

--- a/crates/zccache/src/daemon/server/handle_compile/pipeline.rs
+++ b/crates/zccache/src/daemon/server/handle_compile/pipeline.rs
@@ -238,9 +238,14 @@ pub(super) async fn handle_compile_request(req: CompileRequest<'_>) -> Response 
     } else {
         crate::compiler::SourceMode::Normal
     };
-    let worktree_salt = if worktree_root.is_some()
-        && requires_worktree_in_key(compilation.family, source_mode_for_key)
-    {
+    // Issue #489: PCH/MSVC artifacts embed absolute paths the path-remap
+    // family can't scrub, so the request-level cache must refuse to share
+    // their entries across worktrees regardless of how root-relative the
+    // captured paths look. `requires_worktree_in_key` is the single source
+    // of truth — we mirror it here into both the context-key salt and the
+    // request-cache `worktree_bound` flag so the two cache layers agree.
+    let worktree_bound = requires_worktree_in_key(compilation.family, source_mode_for_key);
+    let worktree_salt = if worktree_root.is_some() && worktree_bound {
         Some(default_key_root.as_path())
     } else {
         None
@@ -337,6 +342,7 @@ pub(super) async fn handle_compile_request(req: CompileRequest<'_>) -> Response 
         client_env: client_env.as_deref(),
         is_rustc,
         worktree_equivalent_context,
+        worktree_bound,
         compile_start,
         parse_args_ns,
         build_context_ns,
@@ -545,6 +551,7 @@ pub(super) async fn handle_compile_request(req: CompileRequest<'_>) -> Response 
                 client_env: client_env.as_deref(),
                 is_rustc,
                 worktree_equivalent_context,
+                worktree_bound,
                 compile_start,
                 parse_args_ns,
                 build_context_ns,

--- a/crates/zccache/src/daemon/server/keys.rs
+++ b/crates/zccache/src/daemon/server/keys.rs
@@ -454,12 +454,22 @@ pub(super) fn request_cache_input_paths(
     paths
 }
 
+/// Build a request-cache entry.
+///
+/// `worktree_bound` must be `true` for any artifact whose bytes embed
+/// absolute paths that no remap pass scrubs — PCH/GCH (`-include`-baked
+/// header AST) and MSVC compiles (PDB/`/Fp`/`/Fd` paths). Setting it forces
+/// `cross_root_shareable = false`, so the entry only matches lookups from
+/// its originating worktree even when every captured path is technically
+/// root-relative. See [`requires_worktree_in_key`] for the truth table and
+/// issue #489 for the diagnostic / DLL-load failure that motivated this.
 pub(super) fn request_cache_entry(
     context_key: ContextKey,
     source_path: &NormalizedPath,
     output_path: &NormalizedPath,
     input_paths: Vec<NormalizedPath>,
     key_root: Option<&NormalizedPath>,
+    worktree_bound: bool,
 ) -> RequestCacheEntry {
     let root = key_root.cloned();
     let root_path = key_root.map(|root| root.as_path());
@@ -469,7 +479,8 @@ pub(super) fn request_cache_entry(
         .iter()
         .map(|path| CachedRequestPath::capture(path, root_path))
         .collect();
-    let cross_root_shareable = root.is_some()
+    let cross_root_shareable = !worktree_bound
+        && root.is_some()
         && source_path.is_root_relative()
         && output_path.is_root_relative()
         && input_paths.iter().all(CachedRequestPath::is_root_relative);

--- a/crates/zccache/src/daemon/server/tests/cache_trim.rs
+++ b/crates/zccache/src/daemon/server/tests/cache_trim.rs
@@ -212,6 +212,7 @@ fn request_cache_resolved_inputs_requires_cross_root_shareable_entry() {
         &output_a,
         vec![source_a.clone(), header_a],
         Some(&NormalizedPath::new(&root_a)),
+        false,
     );
 
     let resolved = request_cache_resolved_inputs(&entry, &NormalizedPath::new(&root_b)).unwrap();
@@ -223,6 +224,69 @@ fn request_cache_resolved_inputs_requires_cross_root_shareable_entry() {
             NormalizedPath::new(root_b.join("include/common.h")),
         ]
     );
+}
+
+// Issue #489: PCH (and MSVC) artifacts bake absolute include paths into the
+// compiled output. `path_remap=off` does not change that — there is no
+// `-ffile-prefix-map` family of scrubbers that touches the PCH AST table. So
+// even when every captured path looks root-relative, a request-level cache
+// entry tagged `worktree_bound` must NEVER be served across worktrees, or the
+// HIT serves a PCH whose embedded paths reference the original worktree.
+//
+// Before the fix, `cross_root_shareable` was derived solely from "are all
+// paths root-relative?" — which is `true` for PCH (the `.h` lives inside the
+// worktree), so a stale fastled10 PCH leaked into a fastled7 build and
+// produced compiler diagnostics referencing `fastled10` paths + downstream
+// DLL load failures (Windows error 126).
+#[test]
+fn request_cache_entry_worktree_bound_rejects_cross_worktree_match() {
+    let tmp = tempfile::tempdir().unwrap();
+    let root_a = NormalizedPath::new(tmp.path().join("fastled10"));
+    let root_b = NormalizedPath::new(tmp.path().join("fastled7"));
+    let source_a: NormalizedPath = root_a.as_path().join("src/FastLED.h").into();
+    let output_a: NormalizedPath = root_a
+        .as_path()
+        .join(".build/meson-quick/ci/meson/native/FastLED.h.pch")
+        .into();
+
+    let pch_entry = request_cache_entry(
+        test_context_key("src/FastLED.h"),
+        &source_a,
+        &output_a,
+        vec![source_a.clone()],
+        Some(&root_a),
+        true,
+    );
+
+    // Same-root lookup: still HITs (sanity).
+    assert!(request_cache_entry_matches_root(&pch_entry, Some(&root_a)));
+
+    // Cross-root lookup: MUST NOT hit, even though every captured path was
+    // root-relative when the entry was built.
+    assert!(
+        !request_cache_entry_matches_root(&pch_entry, Some(&root_b)),
+        "PCH entries are worktree-bound and must not match across worktrees",
+    );
+    assert!(
+        !pch_entry.cross_root_shareable,
+        "worktree-bound entries must opt out of cross-root sharing entirely",
+    );
+
+    // Sanity contrast: an otherwise-identical NON-PCH entry (`worktree_bound = false`)
+    // remains cross-root-shareable. The flag is the single discriminator.
+    let object_entry = request_cache_entry(
+        test_context_key("src/main.cc"),
+        &source_a,
+        &output_a,
+        vec![source_a.clone()],
+        Some(&root_a),
+        false,
+    );
+    assert!(request_cache_entry_matches_root(
+        &object_entry,
+        Some(&root_b)
+    ));
+    assert!(object_entry.cross_root_shareable);
 }
 
 #[test]

